controlplane/kubeadm/controllers/controller.go

```diff
@@ -255,11 +268,23 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 	}
 
 	controlPlane := internal.NewControlPlane(cluster, kcp, ownedMachines)
-	requireUpgrade := controlPlane.MachinesNeedingUpgrade()
+
+	// Aggregate the operational state of all the machines; while aggregating we are adding the
+	// source ref to the aggregate reason (reason@machine/name) so the problem can be easily tracked down to its source.
+	conditions.SetAggregateCondition(controlPlane.KCP, controlplanev1.MachinesReadyCondition, ownedMachines.ConditionGetters(), conditions.AddSourceRef())
+
 	// Upgrade takes precedence over other operations
-	if len(requireUpgrade) > 0 {
+	requireUpgrade := controlPlane.MachinesNeedingUpgrade()
+	switch {
+	case len(requireUpgrade) > 0:
 		logger.Info("Upgrading Control Plane")
+		conditions.MarkFalse(controlPlane.KCP, controlplanev1.ControlPlaneUpgradingCondition, controlplanev1.UpgradingReason, clusterv1.ConditionSeverityWarning, "Upgrading to %s version (%d new version, %d old)", kcp.Spec.Version, kcp.Status.UpdatedReplicas, len(requireUpgrade))
 		return r.upgradeControlPlane(ctx, cluster, kcp, controlPlane)
+	default:
+		// make sure last upgrade operation is marked as completed (if any)
+		if conditions.Has(controlPlane.KCP, controlplanev1.ControlPlaneUpgradingCondition) {
+			conditions.MarkTrue(controlPlane.KCP, controlplanev1.ControlPlaneUpgradingCondition)
+		}
```

Comment:
This feels a bit wrong. It doesn't seem like this is detecting success of upgrade here, but rather blindly setting the condition to true if present if we are not attempting an upgrade.

For example, if someone bumped the kubernetes version on their KCP, the new instance failed to come up correctly and they then reverted the change and deleted the broken new instance, this appears that it would mark the condition as if it were a successful upgrade.

Would it be better to track something like `HasRequestedConfig` (or possibly a better name) as the condition instead? That way the current logic would match the behavioral semantics of how we are setting the condition.

Reply:
Makes sense, I got a little lost while reading the 3rd paragraph, but the problem brought up here can indeed happen. Without storing the wanted version, and the fact that the user has requested an upgrade, it'd be hard to reconcile this condition 🤔

Reply:
_(not for today, otherwise we'll never finish this work)_ The pattern I've seen so far is that we want to set some conditions based by inspecting or performing operations, I was thinking in the future we could associate a function to each condition that performs an action when we need to evaluate it.

Reply:
As agreed, we are removing from this PR the ControlPlaneUpgradingCondition